File: Python/7/example7.py
#!/usr/bin/env python

def do_something():
    print("I'm doing something")


def do_something_with(num):
    print(num * 2)


def do_something_and_return(num):
    return num * 2


def do_something_with_options(num, num2=None):
    if num2:
        return num * num2
    else:
        return num * 2


def do_something_with_opts_plus(num, num2=None, *args):
    my_list = [num,]
    if num2:
        my_list.append(num2)
        if args:
            my_list.extend(args)
    return my_list


def do_something_plus_kw(num, num2=None, *args, **kwargs):
    my_list = do_something_with_opts_plus(num, num2, *args)
    if kwargs:
        for k, v in kwargs.items():
            my_list.append(v)
    return my_list


if __name__ == '__main__':

    print('\nCalling do_something ...')
    do_something()

    print('\nCalling do_something_with using 7 ...')
    do_something_with(7)

    print('\nCalling do_something_and_return using 7 ...')
    val = do_something_and_return(7)
    print(val)

    print('\nCalling do_something_with_options using 7 and 4 ...')
    print(do_something_with_options(7, num2=4))
    print(do_something_with_options(7, 4))
    print(do_something_with_options(7))

    print('\nCalling do_something_with_opts_plus ...')
    print(do_something_with_opts_plus(7, 4, 2, 3, 5))

    print('\nCalling do_something_plus_kw ...')
    print(do_something_plus_kw(7, 4, 2, 3, 5, lucky=7, unlucky=13))

    print('\nAnother spin on do_something_with_opts_plus ...')
    nums=[2, 3, 5]
    print(do_something_with_opts_plus(7, 4, *nums))

    print('\nAnother spin on do_something_plus_kw ...')
    extras={'lucky': 7, 'unluck': 13}
    print(do_something_plus_kw(7, 4, *nums, **extras))

File: Ruby/7/example7.rb
#!/usr/bin/env ruby2

def do_something
    puts "I am doing something."
end


def do_something_with num
    puts "I am doing something with #{num}."
end


def do_something_and_return num
    return num * 2
end


def do_something_with_option num, num2=nil
    return num * num2 if num2 else return num * 2
end


def do_something_with_opts_plus num, *args
    my_list = [num]
    my_list.concat args if args
    return my_list
end


def do_something_plus_kw num, *args, **opts
    my_list = do_something_with_opts_plus num, *args

    if opts
        my_list.push opts[:lucky]
        my_list.push opts[:unlucky]
    end

    return my_list
end


if __FILE__ == $0

    puts "Calling do_something() ..."
    do_something

    puts "Calling do_something_with() ..."
    do_something_with 7

    puts "Calling do_something_and_return using 7 ..."
    puts do_something_and_return 7

    puts "Calling do_something_with_option using 7 ..."
    puts do_something_with_option 7

    puts "Calling do_something_with_option using 7 and 3 ..."
    puts do_something_with_option 7, 3

    puts "Calling do_something_with_opts_plus using 7, 3, 5, 1"
    list = do_something_with_opts_plus 7, 5, 1, num2=3
    puts list

    puts "Calling do_something_with_opts_plus using 7, 5, 1"
    list = do_something_with_opts_plus 7, 5, 1
    puts list

    puts "Calling do_something_plus_kw using 7, 5, 1 and lucky ..."
    puts do_something_plus_kw 7, 5, 1, lucky:7, unlucky:13
end